File: LeetCode/src/greedy/PartitionLabels.java
package greedy;

import java.util.ArrayList;
import java.util.List;

public class PartitionLabels {
    public List<Integer> partitionLabels(String s) {
        int[] last = new int[26];
        for (int i = 0; i < s.length(); i++) {
            last[s.charAt(i) - 'a'] = i;
        }

        List<Integer> result = new ArrayList<>();
        int start = 0, end = 0;
        for (int i = 0; i < s.length(); i++) {
            end = Math.max(end, last[s.charAt(i) - 'a']);
            if (i == end) {
                result.add(end - start + 1);
                start = i + 1;
            }
        }

        return result;
    }

    public static void main(String[] args) {
        PartitionLabels solution = new PartitionLabels();

        String s1 = "ababcbacadefegdehijhklij";
        System.out.println(solution.partitionLabels(s1)); // Output: [9, 7, 8]

        String s2 = "eccbbbbdec";
        System.out.println(solution.partitionLabels(s2)); // Output: [10]
    }
}

File: LeetCode_Python/greedy/PartitionLabels.py
class PartitionLabels:
    def partitionLabels(self, s: str) -> list[int]:
        positions = {char: i for i, char in enumerate(s)}
        res = []
        end = point = 0

        for start, char in enumerate(s):
            end = max(end, positions[char])
            if start == end:
                res.append(end - point + 1)
                point = end + 1

        return res


if __name__ == "__main__":
    solution = PartitionLabels()

    print("Expected Output : [9, 7, 8]")
    print("Actual Output   :", solution.partitionLabels("ababcbacadefegdehijhklij"))

    print("Expected Output : [10]")
    print("Actual Output   :", solution.partitionLabels("eccbbbbdec"))
